Add Prettier with ESLint integration

renchap opened the pull request:
This PR solves #1588 

I made 2 different commits:
The first one:
- Adds [Prettier](https://prettier.io)
- Adds [eslint-plugin-prettier](https://github.com/prettier/eslint-plugin-prettier) to have ESLint throw errors if some files are not formatted using Prettier
- Adds [eslint-config-prettier](https://github.com/prettier/eslint-config-prettier) to disable ESLint rules that conflict with prettier
- Adds a basic Prettier configuration in `package.json`. Prettier recommends to not vary from the default (thats the point of using it), but I added these two as they are more consistent with the current style. They can be removed later if we want to standardise.
- Adds Prettier integration in ESLint configuration
- Removes specific ESLint rules that conflict with Prettier (according to `eslint-config-prettier`).

The second commit is the result of `yarn lint --fix`, effectively running Prettier on all the ESLint-enabled files.

As a result, all tests pass except one (I am not sure where is comes from), and ESLint is all green. As Prettier is ran by ESLint, it will ensure all further PRs will have ran through Prettier.

@ianstormtaylor how does this look to you?

zhujinxuan commented:
BTW, do not forgot to pull from master before eslint.  Otherwise the git may have some unwise delete in the newly merged commit.

renchap commented:
I made a mistake and based this PR from an earlier branch, I will update it in a few minutes :)

zhujinxuan commented:
I am a bit surprised that `eslint --fix` automatically fix the all warnings.

I did that on another project, but manually fixes many problems

renchap commented:
@zhujinxuan when using `eslint-prettier-plugin`, `eslint --fix` will run Prettier on the files. This is a promise of Prettier to not break your code when you pass it through Prettier, so it should not break anything.
The ESLint config changes I made ensured that all the formatting rules are handled by Prettier, and the other rules by ESLint, thus not creating conflicting errors.

zhujinxuan commented:
Can you use ./node_modules/.bin/eslint instead of eslint in the package.json?  Because the eslint use only the global package if it is from globally installed package.

renchap commented on .eslintrc:
```diff
@@ -136,19 +134,12 @@
-    "spaced-comment": ["error", "always", { "exceptions": ["-"]}],
-    "template-curly-spacing": "error",
-    "template-tag-spacing": ["error", "never"],
-    "unicode-bom": ["error", "never"],
```
Yes. Prettier will take care of all the formatting (as it will reformat everything), so these rules are no longer needed.

https://github.com/prettier/eslint-config-prettier disables all those rules, but as its loaded using `extends` all the rules in `.eslintrc` are overriding it. So I went over all the rules and removed the ones specified in `eslint-config-prettier` as useless when using Prettier.

zhujinxuan replied:
@ianstormtaylor Because eslint is conflicting with some styling rules.  I remember eslint-config-prettier has a command to check conflicting rules

renchap commented:
@zhujinxuan `yarn` prepends `./node_modules/.bin/` to the path when running scripts defined in `package.json` (https://docs.npmjs.com/misc/scripts#path)

renchap commented:
@ianstormtaylor You need to have an editor with Prettier support for it to be ran on save (there are plugins for all the major ones: https://prettier.io/docs/en/editors.html).

We can add a script to run Prettier over the whole codebase for those who dont have a Prettier-enabled editor, so they can run it before commit to fix any issues.
The ESLint integration makes ESLint run Prettier on all input files, and throw errors if the file is not the same as the Prettier output, thus ensuring all files processed by ESLint have been formatted using Prettier. Running `yarn lint --fix` will make ESLint run Prettier and save the changes, fixing all issues.

I will add a `.prettierignore` file for `package.json`, as it will be changed by both Yarn and Prettier (and those dont agree on the correct format).

zhujinxuan commented:
@ianstormtaylor I think a common practice is to let editor eslint plugins to run `eslint --fix` after writing files, and reload file after `eslint --fix`

BTW, if you are using Spacevim, you can just 
```
call SpaceVim#layers#load('lang#javascript', 
            \ {
            \ 'auto_fix' : 1,
            \ 'use_local_eslint': 1,
            \ }
            \ )
```

zhujinxuan commented:
@renchap I see there is an unpassed test in the checks.  Just wondering that is due to a bug like `''==null` becames `''===null`.

zhujinxuan commented:
@renchap 

Please manually fix this bug in `./packages/slate/test/schema/custom/node-is-void-invalid-default.js`

The original master set the output as:
```
export const output = (
  <value>
    <document>
      <paragraph>
        {' '}
      </paragraph>
    </document>
  </value>
)
```
The prettier deleted the `{' '}` in the middle

ianstormtaylor commented:
@renchap ah, thanks for the info! Adding a `prettier` script sounds good to me, just to make it super clear—even though it'll just be `"prettier": "yarn run lint --fix"` it sounds like.

@zhujinxuan thank you! That makes sense. Having the editor run ESLint sounds smart.

renchap commented:
I updated the PR:
- Rebased against `master` (as this PR touches a lot of files, it would be best if merged quickly, and it needs to be rebased against master as well `eslint --fix` before merging).
- Added a `.prettierignore` file ignoring files non manually edited
- Fixed the failing test by telling Prettier to not format this line
- Added a `yarn prettier` script, which runs Prettier on all JS(X) files and format them
- Added `prettier --list-different` to `yarn lint` script, so Prettier will check all JS(X) files are formatted correctly, even the ones not handled by ESLint.

As Prettier supports other languages than JS, I would like to have Prettier run with a wider glob: `"**/*.{js,jsx,md,json,css}"`. This is an easy change and can be done later, so we can review these changes in another PR.

majelbstoat commented:
If and when this is merged, setting up a husky/lint-staged rule on pre-commit will prevent unformatted code from ever getting into the codebase.

```
  "scripts": {
    "precommit": "lint-staged"
  },
  "lint-staged": {
    "*.js": ["eslint --fix", "git add"]
  },
```

I have this set up on personal projects, along with a vscode prettier integration, and it's delightful. Fixes on save, guardrail on commit if I do a fix-up in vim.

zhujinxuan commented:
@renchap Perhaps we should use `eslint --fix` in `yarn prettier` command.  The `prettier` command, under some config, cannot fix styling errors detected by eslint.

renchap commented:
@zhujinxuan As per my last commit, Prettier manages other JS files than the ones processed by ESLint (see the changes in the last commit and my previous message about the other formats).

If we want a simple script to fix everything at once, there can be a `yarn autofix` script, which does `yarn lint --fix && yarn prettier`. As there is currently no script to run `eslint --fix`, I did not saw the need to do it.

@majelbstoat yes we can definitely add this, but it might be better discussed in another issue/PR as its not directly dependent on Prettier.

renchap commented:
Updated the PR:
- rebased with `master`
- changed `trailingComma` to `es5`, which will be the new Prettier standard soon and is more consistent with the current codebase.

zhujinxuan commented:
Hi, just wondering if you have `no-continue` in the config.  I am a bit worrying about it is that, because the slate does not allow many helper files in the base, it is uneasy to write helper functions recursively calling each other as a rule combinator.  A method to avoid helper functions is to use while-continue.

For example, I use while-continue for rules combinator in 
https://github.com/ianstormtaylor/slate/pull/1553

renchap commented:
@zhujinxuan I would prefer this PR to only be about Prettier. If you want to suggest/discuss an eslint option, please open a new issue.

@ianstormtaylor Rebasing this PR each time the `master` branch changes is a bit long and painful. Do you want anything to be changed / clarified? I will rebase it once you give me the go.

ianstormtaylor commented:
Okay @renchap, thanks for bearing with me. I've just merged a handful of other pull requests that were ready to go before we change the whole codebase 😄  but if you rebase now I'll merge!

renchap commented:
Dont merge yet, I rebased too quickly :)

renchap commented:
Rebased with the latest `master`, build, lint and tests all green (locally), good to merge as soon as CI agrees @ianstormtaylor 🚀

ianstormtaylor commented:
Thank you very, very much @renchap!

ianstormtaylor commented:
Hey @renchap after merging this, I've pulled and run...

```
yarn clean
yarn install
yarn watch (separate process)
yarn lint
```

...but the linter fails inexplicably with...

```
$ yarn lint
yarn run v1.3.2
$ eslint packages/*/src packages/*/test examples/*/*.js examples/dev/*/*.js && prettier --list-different '**/*.{js,jsx}'
examples/build.dev.js
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

Should that file be ignored? Or am I doing something wrong?

renchap commented:
Yes, `.prettierignore` should be updated, I forgot to test with the dev build.
I see you did it on `master` 👍
